### ISSUE TITLE
OpenShift path based routes with TLS not handled correctly

### ISSUE THREAD

**Issue by mhermsdorfer** [F5Networks/k8s-bigip-ctlr#790]
#### Description
OpenShift routes for different paths with same host incorrectly attempts to apply multiple client-ssl certs with the same cert & SNI name.

If you have certificates & keys defined in multiple routes that have the same host but different ports, k8s-bigip-ctrl attempts to create multiple client-ssl profiles one per route, with the same server-name field, and apply them to the virtual server.  This generates a tmos validation error as you can't have multiple client-ssl profiles with the same cert/sni name.

k8s-bigip-ctrl should be smart enough to see that both routes are for the same host, and since there would be no way to have different certs/profiles for the same host it should re-use the existing cert/key/profile if it exists already from another route.

#### OpenShift Version
3.11

#### Controller Version
1.7

#### BIG-IP Version
13.1.1.0.8

#### Diagnostic Information

```
$ oc get routes -n test-web
NAME                            HOST/PORT                 PATH      SERVICES   PORT       TERMINATION   WILDCARD
mysite123.example.com-default   mysite123.example.com               nginx-ex   8080-tcp   edge/Allow    None
mysite123.example.com-foo       mysite123.example.com     /foo      nginx-ex   8080-tcp   edge/Allow    None
$
```

First route's details:
```
$ oc get routes -n test-web -o yaml mysite123.example.com-default
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  creationTimestamp: 2018-11-07T21:25:23Z
  labels:
    f5type: lab-sf
  name: mysite123.example.com-default
  namespace: test-web
  resourceVersion: "1373937"
  selfLink: /apis/route.openshift.io/v1/namespaces/test-web/routes/mysite123.example.com-default
  uid: a2af9360-e2d3-11e8-b024-005056b79e9e
spec:
  host: mysite123.example.com
  port:
    targetPort: 8080-tcp
  tls:
    certificate: |
      -----BEGIN CERTIFICATE-----
      ... redacted ...
      -----END CERTIFICATE-----
    insecureEdgeTerminationPolicy: Allow
    key: |
      -----BEGIN PRIVATE KEY-----

      ... redacted ...
      -----END PRIVATE KEY-----
    termination: edge
  to:
    kind: Service
    name: nginx-ex
    weight: 100
  wildcardPolicy: None
status:
  ingress: null

```

Then you create a 2nd route, in this case i'm pointing to the same servce for simplicyt, but tykpically it'd be pointed at a different service/port:
```
$ oc get routes -n test-web -o yaml mysite123.example.com-foo
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  creationTimestamp: 2018-11-08T15:56:32Z
  labels:
    f5type: lab-sf
  name: mysite123.example.com-foo
  namespace: test-web
  resourceVersion: "1583483"
  selfLink: /apis/route.openshift.io/v1/namespaces/test-web/routes/mysite123.example.com-foo
  uid: dced97e8-e36e-11e8-9931-005056b73335
spec:
  host: mysite123.example.com
  path: /foo
  port:
    targetPort: 8080-tcp
  tls:
    certificate: |
      -----BEGIN CERTIFICATE-----
      ... redacted ...
      -----END CERTIFICATE-----
    insecureEdgeTerminationPolicy: Allow
    key: |
      -----BEGIN PRIVATE KEY-----
      ... redacted ...
      -----END PRIVATE KEY-----
    termination: edge
  to:
    kind: Service
    name: nginx-ex
    weight: 100
  wildcardPolicy: None
status:
  ingress: null
```

After creating the 2nd route, the BIG-IP will start logging errors in /var/log/ltm:
```
Nov  8 10:56:51 bigip err mcpd[5020]: 0107149e:3: Virtual server /openshift-lab/https-ose-vserver has more than one clientssl/serverssl profile with same server name.
Nov  8 10:56:53 bigip err mcpd[5020]: 0107149e:3: Virtual server /openshift-lab/https-ose-vserver has more than one clientssl/serverssl profile with same server name.
```

If you look at the audit logs, you can see the controller is trying to apply multiple Client-SSL profiles for the same host-name:
```
Nov  8 11:03:28 bigip notice icrd_child[18835]: 01420002:5: AUDIT - pid=18835 user=openshift-lab folder=/openshift-lab module=(tmos)# status=[0107149e:3: Virtual server /openshift-lab/https-ose-vserver has more than one clientssl/serverssl profile with same server name.] cmd_data=modify ltm virtual /openshift-lab/https-ose-vserver { vlans-disabled mask 255.255.255.255 service-down-immediate-action none source-port preserve connection-limit 0 mirror disabled address-status yes destination /openshift-lab/10.71.10.22%0:443 source 0.0.0.0%0/0 translate-port enabled auto-lasthop default metadata replace-all-with { user_agent { value k8s-bigip-ctlr-1.7.0-n1260-443736128 persist true } } mobile-app-tunnel disabled rate-limit-dst-mask 0 nat64 disabled source-address-translation { type automap } gtm-score 0 ip-protocol tcp vlans replace-all-with { } rules { } translate-address enabled enabled cmp-enabled yes profiles replace-all-with { /openshift-lab/default-route-clientssl { context clientside } /Common/http { context all } /openshift-lab/openshift_route_test-web_mysite123.example.com-default-client-ssl { context clientside } /openshift-lab/openshift_route_test-web_mysite123.example.com-foo-client-ssl { context clientside } /Common/tcp { context all } } rate-limit disabled policies replace-all-with { /openshift-lab/openshift_secure_routes } rate-limit-mode object rate-limit-src-mask 0 app-service none auth replace-all-with { } bwc-policy none description none fallback-persistence none flow-eviction-policy none last-hop-pool none per-flow-request-access-policy none pool none rate-class none related-rules { } security-log-profiles replace-all-with { } service-policy none traffic-classes replace-all-with { } transparent-nexthop none urldb-feed-policy none }
```

A work around for this issue is to pre-configure the certificates, keys and ssl profiles on the BIG-IP out of band of the OpenShift infrastructure.  You can then reference the pre-configured profiles in the OpenShift route annotations.  However, this is problematic since application teams can not manage their certificates inside of OpenShift.

**Comment by vincentmli** [comment]
the problem here is that when using 2  route with different route name but with same host name. the CC would create two clientssl profile with the route name as the clientssl profile name as code shown below 
<pre>

		// We process the profile from the Route
		if "" != route.Spec.TLS.Certificate && "" != route.Spec.TLS.Key {
			profile := makeRouteClientSSLProfileRef(
				rsCfg.Virtual.Partition, sKey.Namespace, route.ObjectMeta.Name) <===HERE

			cp := NewCustomProfile(
				profile,
				route.Spec.TLS.Certificate,
				route.Spec.TLS.Key,
				route.Spec.Host,
				false,
				"", // peerCertMode
				"", // caFile
			)
</pre>

since the CC will for loop each route ( say route A and B) and create different clientssl profile with route name be part of the clientssl profile name, and when different route use same cert/key literal information, this result in 

openshift_route_\<namespace\>\_\<routeA\>-client-ssl 
openshift_route_\<namespace\>\_\<routeB\>-client-ssl

with same cert/key/hostname

maybe the CC should create clientssl profile name with hostname instead of route name, not to create duplicate clientssl profile with same hostname/cert/key.

**Comment by vincentmli** [comment]
I tried to re-create the issue in lab and I get better idea now, it is not the problem of two duplicate clientssl profile with same cert/key/hostname, it is the problem that BIGIP/mcpd does not accept two clientssl profile with same "server-name" to be applied to https-ose-vserver, in CC when creating the clientssl profile, 

<pre>
		// We process the profile from the Route
		if "" != route.Spec.TLS.Certificate && "" != route.Spec.TLS.Key {
			profile := makeRouteClientSSLProfileRef(
				rsCfg.Virtual.Partition, sKey.Namespace, route.ObjectMeta.Name) 

			cp := NewCustomProfile(
				profile,
				route.Spec.TLS.Certificate,
				route.Spec.TLS.Key,
				route.Spec.Host, <====this applied to server-name
				false, <=SNI bool setting, if SNI false here, no need to set server-name in clientssl
				"", // peerCertMode
				"", // caFile
			)

func NewCustomProfile(
	profile ProfileRef,
	cert,
	key,
	serverName string,
	sni bool,
	peerCertMode,
	caFile string,
) CustomProfile {
	cp := CustomProfile{
		Name:         profile.Name,
		Partition:    profile.Partition,
		Context:      profile.Context,
		Cert:         cert,
		Key:          key,
		ServerName:   serverName, <======server-name
		SNIDefault:   sni, <=====SNI
		PeerCertMode: peerCertMode,
	}
	if peerCertMode == peerCertRequired {
		cp.CAFile = caFile
	}
	return cp
}

</pre>

**Comment by vincentmli** [comment]
here is the one-line change to fix the issue:

<pre>

diff --git a/pkg/appmanager/profiles.go b/pkg/appmanager/profiles.go
index 74266ed..49913a5 100644
--- a/pkg/appmanager/profiles.go
+++ b/pkg/appmanager/profiles.go
@@ -109,7 +109,7 @@ func (appMgr *Manager) setClientSslProfile(
                                profile,
                                route.Spec.TLS.Certificate,
                                route.Spec.TLS.Key,
-                               route.Spec.Host,
+                               "",
                                false,
                                "", // peerCertMode
                                "", // caFile
</pre>

two test route yaml

<pre>
apiVersion: v1
kind: Route
metadata:
  name: route1
  labels:
    # Tells the BIG-IP Controller to watch Routes with a specific label.
    # Include the label in the k8s-bigip-ctlr Deployment under "args"
    # (for example, "--route-label="App1")
    f5type: App1
  annotations:
    virtual-server.f5.com/balance: least-connections-member
    virtual-server.f5.com/health: |
        [
         {
              "path":     "www.example.com/",
              "send":     "GET / HTTP/1.1\\r\\nHost: www.example.com\\r\\nConnection: Close\\r\\n\\r\\n",
              "type": "http",
              "interval": 5,
              "timeout": 16
          }
        ]
spec:
  host: www.example.com
  to:
     kind: Service
     name: nginxservice
  tls:
     termination: edge
     insecureEdgeTerminationPolicy: Allow
     certificate:  |
        -----BEGIN CERTIFICATE-----
        ...............
        -----END CERTIFICATE-----
     key:  |
        -----BEGIN PRIVATE KEY-----
        ...............
        -----END PRIVATE KEY-----

apiVersion: v1
kind: Route
metadata:
  name: route2
  labels:
    # Tells the BIG-IP Controller to watch Routes with a specific label.
    # Include the label in the k8s-bigip-ctlr Deployment under "args"
    # (for example, "--route-label="App1")
    f5type: App1
  annotations:
    virtual-server.f5.com/balance: least-connections-member
    virtual-server.f5.com/health: |
        [
         {
              "path":     "www.example.com/foo",
              "send":     "GET / HTTP/1.1\\r\\nHost: www.example.com\\r\\nConnection: Close\\r\\n\\r\\n",
              "type": "http",
              "interval": 5,
              "timeout": 16
          }
        ]
spec:
  host: www.example.com
  path: "/foo"
  to:
     kind: Service
     name: nginxservice
  tls:
     termination: edge
     insecureEdgeTerminationPolicy: Allow
     certificate:  |
        -----BEGIN CERTIFICATE-----
        ...............
        -----END CERTIFICATE-----
     key:  |
        -----BEGIN PRIVATE KEY-----
        ...............
        -----END PRIVATE KEY-----
</pre>

result in BIGIP:

<pre>
ltm profile client-ssl openshift_route_default_route1-client-ssl {
    app-service none
    cert openshift_route_default_route1-client-ssl.crt
    cert-key-chain {
        openshift_route_default_route1-client-ssl {
            cert openshift_route_default_route1-client-ssl.crt
            key openshift_route_default_route1-client-ssl.key
        }
    }
    chain none
    defaults-from /Common/clientssl
    inherit-certkeychain false
    key openshift_route_default_route1-client-ssl.key
    passphrase none
    server-name none <====none
    sni-default false <====false
}
ltm profile client-ssl openshift_route_default_route2-client-ssl {
    app-service none
    cert openshift_route_default_route2-client-ssl.crt
    cert-key-chain {
        openshift_route_default_route2-client-ssl {
            cert openshift_route_default_route2-client-ssl.crt
            key openshift_route_default_route2-client-ssl.key
        }
    }
    chain none
    defaults-from /Common/clientssl
    inherit-certkeychain false
    key openshift_route_default_route2-client-ssl.key
    passphrase none
    server-name none . <====none
    sni-default false <====false
}

ltm virtual https-ose-vserver {
    destination 192.168.2.136%0:https
    ip-protocol tcp
    mask 255.255.255.255
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.8.0-WIP-vli-cd01ce5-dirty-20181130110437
        }
    }
    partition openshift
    policies {
        openshift_secure_routes { }
    }
    profiles {
        /Common/http { }
        /Common/tcp { }
        default-route-clientssl {
            context clientside
        }
        openshift_route_default_route1-client-ssl { <=====
            context clientside
        }
        openshift_route_default_route2-client-ssl { <=====
            context clientside
        }
    }
    source 0.0.0.0/0
    source-address-translation {
        type automap
    }
    translate-address enabled
    translate-port enabled
    vs-index 72
}

ltm policy openshift_insecure_routes {
    controls { forwarding }
    last-modified 2018-11-30:03:16:40
    partition openshift
    requires { http }
    rules {
        openshift_route_default_route1 {
            actions {
                0 {
                    forward
                    select
                    pool openshift_default_nginxservice
                }
            }
            conditions {
                0 {
                    http-host
                    host
                    values { www.example.com }
                }
            }
            ordinal 1
        }
        openshift_route_default_route2 { <=====
            actions {
                0 {
                    forward
                    select
                    pool openshift_default_nginxservice
                }
            }
            conditions {
                0 {
                    http-host
                    host
                    values { www.example.com }
                }
                1 { <=====
                    http-uri
                    path-segment
                    index 1
                    values { foo } <======
                }
            }
        }
    }
    status legacy
    strategy /Common/first-match
}
</pre>

**Comment by trinaths** [comment]
Issue fixed in 1.11

https://github.com/F5Networks/k8s-bigip-ctlr/blob/master/docs/RELEASE-NOTES.rst
